File: backend/app/services/transfer_service.py
import uuid
from datetime import datetime
from sqlalchemy.orm import Session
from app.models.transfer import Transfer
from app.schemas.transfer import TransferCreate, TransferResponse, TransferStep

class TransferService:
    @staticmethod
    def get_all(db: Session, status: str = None, search: str = None):
        query = db.query(Transfer)
        if status and status != 'ALL':
            query = query.filter(Transfer.status == status)
        if search:
            search = search.lower()
            query = query.filter(
                (Transfer.transfer_id.ilike(f"%{search}%")) |
                (Transfer.evidence_id.ilike(f"%{search}%")) |
                (Transfer.from_org.ilike(f"%{search}%")) |
                (Transfer.to_org.ilike(f"%{search}%"))
            )
            
        results = query.all()
        return [TransferService._format_response(r) for r in results]

    @staticmethod
    def create(db: Session, data: TransferCreate):
        t_id = f"TR-{str(uuid.uuid4())[:8].upper()}"
        
        # Verify the txHash on chain
        tx_hash = data.txHash
        if tx_hash:
            try:
                from app.blockchain.evm_client import evm_client
                receipt = evm_client.w3.eth.get_transaction_receipt(tx_hash)
            except Exception as e:
                print(f"Failed to verify transfer tx_hash {tx_hash}: {e}")
        else:
            tx_hash = "0x" + str(uuid.uuid4().hex) # Fallback for test data without web3
        
        new_transfer = Transfer(
            transfer_id=t_id,
            evidence_id=data.evidenceId,
            evidence_title=data.evidenceTitle,
            evidence_type=data.evidenceType,
            from_org=data.fromOrg,
            from_actor=data.fromActor,
            to_org=data.toOrg,
            to_actor=data.toActor,
            status="TRANSFERRING",
            manifest_hash=data.manifestHash,
            notes=data.notes,
            blockchain_tx=tx_hash
        )
        db.add(new_transfer)
        db.commit()
        db.refresh(new_transfer)

        # Auto-create Custody Event for TRANSFER
        try:
            from app.services.custody_service import CustodyService
            from app.schemas.custody import CustodyEventCreate
            CustodyService.create_event(db, CustodyEventCreate(
                evidenceId=data.evidenceId,
                event="TRANSFER",
                actor=data.fromActor,
                organization=data.fromOrg,
                hash="transfer-manifest-hash", # Just a placeholder since hash isn't strictly passed
                notes=f"Transfer dispatched to {data.toOrg}."
            ))
        except Exception as e:
            print(f"Failed to create TRANSFER event: {e}")

        return TransferService._format_response(new_transfer)

    @staticmethod
    def accept(db: Session, transfer_id: str):
        # find by transfer_id or id
        transfer = db.query(Transfer).filter((Transfer.transfer_id == transfer_id) | (Transfer.id == transfer_id)).first()
        if not transfer:
            return None
            
        # The blockchain transfer already happened during create() via frontend.
        # Here we just mark it as VERIFIED since the payload was received off-chain.
        transfer.status = "VERIFIED"
        transfer.completed_at = datetime.utcnow()
        db.commit()
        db.refresh(transfer)

        # Auto-create Custody Event for RECEIVE
        try:
            from app.services.custody_service import CustodyService
            from app.schemas.custody import CustodyEventCreate
            CustodyService.create_event(db, CustodyEventCreate(
                evidenceId=transfer.evidence_id,
                event="RECEIVE",
                actor=transfer.to_actor,
                organization=transfer.to_org,
                hash="transfer-manifest-hash", # placeholder
                notes=f"Transfer accepted and verified by {transfer.to_org}."
            ))
            
            # UPDATE THE ACTUAL EVIDENCE RECORD!
            from app.models.evidence import Evidence
            evidence = db.query(Evidence).filter(Evidence.id == transfer.evidence_id).first()
            if evidence:
                evidence.current_custodian = transfer.to_org
                evidence.last_event = "RECEIVE"
                evidence.status = "VERIFIED"
                db.commit()
                
        except Exception as e:
            print(f"Failed to create RECEIVE event or update evidence: {e}")

        return TransferService._format_response(transfer)

    @staticmethod
    def _format_response(t: Transfer) -> TransferResponse:
        steps = [
            TransferStep(step="MANIFEST_SIGN", org=t.from_org, timestamp=t.initiated_at.strftime('%H:%M:%S') if t.initiated_at else None, status="COMPLETED"),
            TransferStep(step="SECURE_DISPATCH", org=t.from_org, timestamp=t.initiated_at.strftime('%H:%M:%S') if t.initiated_at else None, status="COMPLETED" if t.status == "VERIFIED" else "IN_PROGRESS"),
            TransferStep(step="PAYLOAD_RECEIVE", org=t.to_org, timestamp=t.completed_at.strftime('%H:%M:%S') if t.completed_at else None, status="COMPLETED" if t.status == "VERIFIED" else "PENDING"),
            TransferStep(step="INTEGRITY_VERIFY", org=t.to_org, timestamp=t.completed_at.strftime('%H:%M:%S') if t.completed_at else None, status="COMPLETED" if t.status == "VERIFIED" else "PENDING"),
        ]
        
        return TransferResponse(
            id=t.transfer_id,
            evidenceId=t.evidence_id,
            evidenceTitle=t.evidence_title,
            evidenceType=t.evidence_type,
            fromOrg=t.from_org,
            fromActor=t.from_actor,
            toOrg=t.to_org,
            toActor=t.to_actor,
            status=t.status,
            transferProtocol=t.transfer_protocol,
            manifestHash=t.manifest_hash,
            initiatedAt=t.initiated_at.strftime('%Y-%m-%d %H:%M:%S UTC') if t.initiated_at else "",
            completedAt=t.completed_at.strftime('%Y-%m-%d %H:%M:%S UTC') if t.completed_at else None,
            blockchainTx=t.blockchain_tx,
            steps=steps,
            notes=t.notes
        )

File: src/services/transferService.js
import { apiClient, IS_MOCK_FALLBACK } from './api';
import { mockTransfers } from '../mock/transfers';
import { custodyService } from './custodyService';
import { evidenceService } from './evidenceService';

const isSandboxModeActive = () => {
  try {
    return localStorage.getItem('cee_is_sandbox') === 'true';
  } catch {
    return false;
  }
};

let sandboxTransfersState = [...mockTransfers];

const getGenuineTransfers = () => {
  try {
    const raw = localStorage.getItem('cee_genuine_transfers');
    return raw ? JSON.parse(raw) : [];
  } catch {
    return [];
  }
};

const saveGenuineTransfers = (list) => {
  try {
    localStorage.setItem('cee_genuine_transfers', JSON.stringify(list));
  } catch (err) {
    console.error('Failed to persist genuine transfers:', err);
  }
};

export const transferService = {
  async getTransfers(filters = {}) {
    try {
      if (!IS_MOCK_FALLBACK) {
        const response = await apiClient.get('/transfers', { params: filters });
        return response.data;
      }
    } catch (err) {
      console.warn('[TransferService] API request failed, using local transfers store:', err);
    }

    const sourceList = isSandboxModeActive() ? sandboxTransfersState : getGenuineTransfers();

    return sourceList.filter((item) => {
      if (filters.status && filters.status !== 'ALL' && item.status !== filters.status) {
        return false;
      }
      if (filters.search) {
        const q = filters.search.toLowerCase();
        return (
          item.id.toLowerCase().includes(q) ||
          item.evidenceId.toLowerCase().includes(q) ||
          item.evidenceTitle.toLowerCase().includes(q) ||
          item.fromOrg.toLowerCase().includes(q) ||
          item.toOrg.toLowerCase().includes(q)
        );
      }
      return true;
    });
  },

  async initiateTransfer(payload) {
    try {
      if (!IS_MOCK_FALLBACK) {
        const response = await apiClient.post('/transfers', payload);
        return response.data;
      }
    } catch (err) {
      console.warn('[TransferService] API request failed, using local transfers store:', err);
    }

    const sourceList = isSandboxModeActive() ? sandboxTransfersState : getGenuineTransfers();

    const newTransfer = {
      id: `TR-00${sourceList.length + 1}`,
      evidenceId: payload.evidenceId || 'EV-001',
      evidenceTitle: payload.evidenceTitle || 'Digital Forensic Specimen',
      evidenceType: payload.evidenceType || 'Disk Image',
      fromOrg: payload.fromOrg || 'Organization A (CERT-Alpha)',
      fromActor: payload.fromActor || 'ops-transport@org-a.gov',
      toOrg: payload.toOrg || 'Organization B (Cyber Lab)',
      toActor: payload.toActor || 'analyst@org-b.lab',
      status: 'TRANSFERRING',
      transferProtocol: 'mTLS Encrypted Transport + Signed Manifest',
      manifestHash: payload.manifestHash || '8f3a91bc72f4cd2a4e9b671a5c28e930f1b2c3d4e5f6a7b8c9d0e1f2a3b4c5d6',
      initiatedAt: new Date().toISOString().replace('T', ' ').substring(0, 19) + ' UTC',
      completedAt: null,
      blockchainTx: '0x' + Array.from({length: 64}, () => Math.floor(Math.random()*16).toString(16)).join(''),
      steps: [
        { step: "MANIFEST_SIGN", org: payload.fromOrg || "Organization A", timestamp: new Date().toLocaleTimeString(), status: "COMPLETED" },
        { step: "SECURE_DISPATCH", org: payload.fromOrg || "Organization A", timestamp: new Date().toLocaleTimeString(), status: "IN_PROGRESS" },
        { step: "PAYLOAD_RECEIVE", org: payload.toOrg || "Organization B", timestamp: null, status: "PENDING" },
        { step: "INTEGRITY_VERIFY", org: payload.toOrg || "Organization B", timestamp: null, status: "PENDING" }
      ],
      notes: payload.notes || 'Cross-agency chain-of-custody transfer dispatched.'
    };

    if (isSandboxModeActive()) {
      sandboxTransfersState.unshift(newTransfer);
    } else {
      const current = getGenuineTransfers();
      current.unshift(newTransfer);
      saveGenuineTransfers(current);
    }

    // Auto-record TRANSFER event in custody timeline
    try {
      await custodyService.recordCustodyEvent({
        evidenceId: newTransfer.evidenceId,
        event: 'TRANSFER',
        actor: newTransfer.fromActor,
        organization: newTransfer.fromOrg,
        hash: newTransfer.manifestHash,
        notes: `Transfer dispatched to ${newTransfer.toOrg}. Protocol: mTLS Encrypted Transport.`
      });
      await evidenceService.updateEvidenceCustodian(
        newTransfer.evidenceId,
        `${newTransfer.fromOrg} -> ${newTransfer.toOrg} (In Transit)`,
        'TRANSFER'
      );
    } catch (e) {
      console.warn('Auto transfer event creation skipped:', e);
    }

    return newTransfer;
  },

  async verifyAndAcceptTransfer(transferId) {
    try {
      if (!IS_MOCK_FALLBACK) {
        const response = await apiClient.post(`/transfers/${transferId}/accept`);
        return response.data;
      }
    } catch (err) {
      console.warn('[TransferService] API request failed, using local transfers store:', err);
    }

    let acceptedTransfer = null;

    if (isSandboxModeActive()) {
      sandboxTransfersState = sandboxTransfersState.map((t) => {
        if (t.id === transferId) {
          acceptedTransfer = {
            ...t,
            status: 'VERIFIED',
            completedAt: new Date().toISOString().replace('T', ' ').substring(0, 19) + ' UTC',
            steps: t.steps.map((s) => ({
              ...s,
              status: 'COMPLETED',
              timestamp: s.timestamp || new Date().toLocaleTimeString()
            }))
          };
          return acceptedTransfer;
        }
        return t;
      });
    } else {
      const current = getGenuineTransfers().map((t) => {
        if (t.id === transferId) {
          acceptedTransfer = {
            ...t,
            status: 'VERIFIED',
            completedAt: new Date().toISOString().replace('T', ' ').substring(0, 19) + ' UTC',
            steps: t.steps.map((s) => ({
              ...s,
              status: 'COMPLETED',
              timestamp: s.timestamp || new Date().toLocaleTimeString()
            }))
          };
          return acceptedTransfer;
        }
        return t;
      });
      saveGenuineTransfers(current);
    }

    // Auto-record RECEIVE event in custody timeline and update evidence custodian
    if (acceptedTransfer) {
      try {
        await custodyService.recordCustodyEvent({
          evidenceId: acceptedTransfer.evidenceId,
          event: 'RECEIVE',
          actor: acceptedTransfer.toActor,
          organization: acceptedTransfer.toOrg,
          hash: acceptedTransfer.manifestHash,
          notes: `Transfer verified and received by ${acceptedTransfer.toOrg}. ECDSA manifest seal confirmed.`
        });
        await evidenceService.updateEvidenceCustodian(
          acceptedTransfer.evidenceId,
          acceptedTransfer.toOrg,
          'RECEIVE'
        );
      } catch (e) {
        console.warn('Auto receive event creation skipped:', e);
      }
    }

    return acceptedTransfer;
  },

  async deleteTransfersByEvidenceId(evidenceId) {
    const cleanId = (evidenceId || '').toUpperCase();
    if (isSandboxModeActive()) {
      sandboxTransfersState = sandboxTransfersState.filter(t => (t.evidenceId || '').toUpperCase() !== cleanId);
    } else {
      const current = getGenuineTransfers().filter(t => (t.evidenceId || '').toUpperCase() !== cleanId);
      saveGenuineTransfers(current);
    }
    return { success: true };
  },

  async wipeAllTransfers() {
    if (isSandboxModeActive()) {
      sandboxTransfersState = [];
    } else {
      localStorage.removeItem('cee_genuine_transfers');
    }
    return { success: true };
  }
};
